server/channels/store/sqlstore/property_field_store.go

```diff
@@ -0,0 +1,339 @@
+}
+
+func (s *SqlPropertyFieldStore) Create(field *model.PropertyField) (*model.PropertyField, error) {
+	if field.ID != "" {
+		return nil, store.NewErrInvalidInput("PropertyField", "id", field.ID)
+	}
+
+	field.PreSave()
+
+	if err := field.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "property_field_create_isvalid")
+	}
+
+	insertMap, err := s.propertyFieldToInsertMap(field)
+	if err != nil {
+		return nil, err
+	}
+
+	queryString, args, err := s.getQueryBuilder().
+		Insert("PropertyFields").
+		SetMap(insertMap).
+		ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "property_field_create_tosql")
+	}
+
+	if _, err := s.GetMaster().Exec(queryString, args...); err != nil {
+		return nil, errors.Wrap(err, "property_field_create_insert")
+	}
+
+	return field, nil
+}
+
+func (s *SqlPropertyFieldStore) Get(id string) (*model.PropertyField, error) {
+	queryString, args, err := s.getQueryBuilder().
+		Select(propertyFieldColumns...).
+		From("PropertyFields").
+		Where(sq.Eq{"id": id}).
+		ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "property_field_get_tosql")
+	}
+
+	rows, err := s.GetReplica().Query(queryString, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "property_field_get_select")
+	}
+	defer rows.Close()
+
+	field, err := propertyFieldFromRows(rows)
+	if err != nil {
+		return nil, errors.Wrap(err, "property_field_get_propertyfieldfromrows")
+	}
+
+	return field, nil
+}
+
+func (s *SqlPropertyFieldStore) GetMany(ids []string) ([]*model.PropertyField, error) {
+	queryString, args, err := s.getQueryBuilder().
+		Select(propertyFieldColumns...).
+		From("PropertyFields").
+		Where(sq.Eq{"id": ids}).
+		ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "property_field_get_many_tosql")
+	}
+
+	rows, err := s.GetReplica().Query(queryString, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "property_field_get_many_query")
+	}
+	defer rows.Close()
+
+	fields, err := propertyFieldsFromRows(rows)
+	if err != nil {
+		return nil, errors.Wrap(err, "property_field_get_many_propertyfieldfromrows")
+	}
+
+	if len(fields) < len(ids) {
+		return nil, errors.New("property_field_get_many_missmatch_results")
+	}
+
+	return fields, nil
+}
+
+func (s *SqlPropertyFieldStore) SearchPropertyFields(opts model.PropertyFieldSearchOpts) ([]*model.PropertyField, error) {
+	if opts.Page < 0 {
+		return nil, errors.New("page must be positive integer")
+	}
+
+	if opts.PerPage < 0 {
+		return nil, errors.New("per page must be positive integer")
+	}
+
+	query := s.getQueryBuilder().
+		Select(propertyFieldColumns...).
+		From("PropertyFields").
+		OrderBy("CreateAt ASC").
+		Offset(uint64(opts.Page * opts.PerPage)).
+		Limit(uint64(opts.PerPage))
+
+	if !opts.IncludeDeleted {
+		query = query.Where(sq.Eq{"DeleteAt": 0})
+	}
+
+	if opts.GroupID != "" {
+		query = query.Where(sq.Eq{"GroupID": opts.GroupID})
+	}
+
+	if opts.TargetType != "" {
+		query = query.Where(sq.Eq{"TargetType": opts.TargetType})
```

Comment:
Note that you only have a unique index on `(GroupID, TargetID, Name)`. This means queries to filter by just DeleteAt or TargetType, or even just TargetID won't use any index. Should we add an index for those? Or are we confident that there won't be too many SQL queries using those fields?

Reply:
Features should use searches that are narrowed down to a specific group at least, if not a group and a target. We can enforce this by changing the signature of the service search method to receive those fields explicitly and add them to the options for good measure, and add more indexes if we start having the need to run different queries that escape this restriction.

What do you think?

Reply:
Ok. Changing the signature of those search methods to use those fields explicitly sounds like a good idea. :+1: